Refactor gempak jobs for new COM and style

WalterKolczynski-NOAA opened the pull request:
# Description
Updates the gempak jobs to fit the new COM structure while also refactoring them some to improve the style. Despite these technical changes, the overall structure is left unchanged for most scripts, though some have been rewritten to make the needed changes easier. Some of these scripts had already been updated some in the original COM refactor and thus needed fewer updates.

Style updates includes converting all gempak scripts to bash, making them shellcheck compliant, and removing trailing whitespace. Further refactoring to improve maintainability will be needed in the future (see #2341, #2342, #2343, #2348).

The GFS gif scripts were identical except the forecast hour, so they are collapsed down into two: one for f000 and one for other forecast hours.

The gempak executables have short path limits. To get around this without having the gempak module recompiled, target directories (mostly relevant for the gempak meta jobs) are symlinked into the working directory to drasticly reduce the path lengths.

Part of this update includes replacing existing MPMD calls with the new standard `ush/run_mpmd.sh` script.

A new function, `wait_for_file()`, is introduced to standardize waiting for a file to be available.

Gempak forecast hours are often hard-coded within scripts. In addition to issues with maintainability, this causes problems for shorter forecasts, such as we typically run for testing purposes. For now, we simply check the values against the forecast length and reduce if necessary. Future work (#2348) will be needed to remove these hard-coded values with variables set in the config file (or just use update gempak products to match standard output time variables).

One-degree gempak files have been updated to include `1p00` in the filename.

Several gempak job dependencies are corrected.

Fake gempak data for external models is being staged on tier-1 machines to allow testing.

**Output has not been verified.** Future PRs will likely be needed to bring full functionality online.

Resolves #2158
Resolves #2152
Resolves #2151
Resolves #2249
Resolves #2247
Refs #2157
Refs #2348 

# Type of change
- Bug fix (fixes something broken)
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- [x] 4-cycle 120-h on Orion
- [x] 33-cycle 384-h on WCOSS2

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

WalterKolczynski-NOAA committed:
Disable warnings for functions in test statements

Disables warning that functions being used in test statements (such
as if, while, etc.) will not capture errors.

WalterKolczynski-NOAA committed:
Ignore shellcheck warning in config.base

Ignores warning about single quotes not expanding in the gfs config.base.
The warning is already ignored for gefs.

WalterKolczynski-NOAA committed:
Update gempak meta_trop for COM refactor

Missed one on previous pass

WalterKolczynski-NOAA committed:
update gempak meta qpf for COM refactor

WalterKolczynski-NOAA committed:
Add missing error check to gempak meta mar ql

WalterKolczynski-NOAA committed:
Fix lookback date calculation in gempak meta sa2

WalterKolczynski-NOAA committed:
Fix gempak meta sa2

WalterKolczynski-NOAA committed:
Fix cleanup job dependencies

WalterKolczynski-NOAA committed:
Cap gempak post frequency to rate produced

WalterKolczynski-NOAA committed:
Update upper-air paths

WalterKolczynski-NOAA committed:
Update table path for GOES gempak

WalterKolczynski-NOAA committed:
Only create symlinks for external gempak if they don't exist

`ln -sf` wasn't overwriting existing links, but we don't need to do
the operation if the symlink already exists anyway.

WalterKolczynski-NOAA committed:
Refactor gempak jobs for new COM and style

Updates the gempak jobs to fit the new COM structure while also
refactoring them some to improve the style. Despite these technical
changes, the overall structure is left unchanged for most scripts,
though some have been rewritten to make the needed changes easier. Some
of these scripts had already been updated some in the original COM
refactor and thus needed fewer updates. Further refactoring to improve
maintainability will be needed in the future.

The GFS gif scripts were identical except the forecast hour, so they are
collapsed down into two: one for f000 and one for other forecast
hours.

One known issue remains that prevent all gempak jobs from running to
completion: `rdbfmsua` is not currently being built by
NOAA-EMC/gfs-utils. The refactor of this program to cmake was defered
due to the difficulty in finding the gempak package. This causes the
`gfsgempakncdcupapgif` job to fail, so #2157 is not fully resolved in
this PR. That line has been temporarily commented out so jobs can
complete.

Another issue is the gempak executables have short path limits. To get
around this without having the gempak module recompiled, target
directories (mostly relevant for the gempak meta jobs) are symlinked
into the working directory to drasticly reduce the path lengths.

Part of this update includes replacing existing MPMD calls with the new
standard `ush/run_mpmd.sh` script.

A new function, `wait_for_file()`, is introduced to standardize waiting
for a file to be available.

Gempak forecast hours are often hard-coded within scripts. In addition
to issues with maintainability, this causes problems for shorter
forecasts, such as we typically run for testing purposes. For now, we
simply check the values against the forecast length and reduce if
necessary. Future work will be needed to remove these hard-coded values
with variables set in the config file (or just use update gempak
products to match standard output time variables).

Several gempak job dependencies are corrected.

Output has not been confirmed to exist or verified. Future PRs will
likely be needed to bring full functionality online.

Resolves #2158
Resolves #2152
Resolves #2151
Resolves #2249
Resolves #2247
Refs #2157

WalterKolczynski-NOAA committed:
Allow non-matching bufr table

PR #2383 added a new bufr table, along with a fatal error if there was
not one that matched the resolution. However, workflow usually uses a
low resolution to test, and it works fine even if the bufr parm file is
a mismatch, so turned the fatal error into a warning and links the bufr
parm file as before.

WalterKolczynski-NOAA committed:
Fix undefined parm variable for bufr

GwenChen-NOAA commented on gempak/ush/gdas_ecmwf_meta_ver.sh:
```diff
@@ -195,28 +136,28 @@ EOFplt
-        ${DBNROOT}/bin/dbn_alert MODEL ECMWFVER_HPCMETAFILE $job \
-        ${COMOUTecmwf}.${PDY}/meta/ecmwfver_${PDY}_${cyc}
-    fi
+mv ecmwfver.meta "${COM_ATMOS_GEMPAK_META}/ecmwfver_${PDY}_${cyc2}"
```
Not sure why this script creates a ecmwf meta file. If ${COM_ATMOS_GEMPAK_META} is pointed to gdas.20240310/00/atmos/gempak/meta, there is no ecmwfver_20240310_00 file under /lfs/h1/ops/prod/com/gfs/v16.3/gdas.20240310/00/atmos/gempak/meta. In the old code, it is sent to ${COMOUTecmwf}.${PDY}/meta. I don't see it under /lfs/h1/ops/prod/com/ecmwf/v2.1/ecmwf.20240310/gempak either.

GwenChen-NOAA replied:
This file is generated for the 06 cycle only. It is located at /lfs/h1/ops/prod/com/gfs/v16.3/gdas.20240315/06/atmos/gempak/meta.20240315/meta.

GwenChen-NOAA commented on gempak/ush/gdas_ukmet_meta_ver.sh:
```diff
@@ -226,22 +139,28 @@ EOFplt
-        ${COMOUTukmet}/ukmet.${PDY}/meta/ukmetver_${PDY}_12
-    fi
+if (( err != 0 )) || [[ ! -s ukmetver_12.meta ]]; then
+    echo "FATAL ERROR: Failed to create ecmwf meta file"
```
The text message should be "FATAL ERROR: Failed to create **ukmet** meta file".

WalterKolczynski-NOAA replied:
Fixed

GwenChen-NOAA commented on gempak/ush/gdas_ukmet_meta_ver.sh:
```diff
@@ -226,22 +139,28 @@ EOFplt
+    exit "${err}"
+fi
+
+mv ukmetver_12.meta "${COM_ATMOS_GEMPAK_META}/ukmetver_${PDY}_12"
```
Similar to the ecmwf meta file, there is no ukmetver_20240310_12 file under /lfs/h1/ops/prod/com/gfs/v16.3/gdas.20240310/00/atmos/gempak/meta if it is where ${COM_ATMOS_GEMPAK_META} pointed to. I don't see it under /lfs/h1/ops/prod/com/ukmet/v2.2/ukmet.20240310/gempak/meta either.

GwenChen-NOAA replied:
This file is generated for the 06 cycle only. It is located at /lfs/h1/ops/prod/com/gfs/v16.3/gdas.20240315/06/atmos/gempak/meta/ukmet.20240315/meta.

GwenChen-NOAA commented on gempak/ush/gdas_ukmet_meta_ver.sh:
```diff
@@ -226,22 +139,28 @@ EOFplt
+
+if [[ "${SENDDBN}" == "YES" ]] ; then
+    "${DBNROOT}/bin/dbn_alert" MODEL UKMETVER_HPCMETAFILE "${job}" \
+        "${COM_GEMPAK_META}/ukmetver_${PDY}_12"
```
This line should be "${COM_**ATMOS**_GEMPAK_META}/ukmetver_${PDY}_12".

WalterKolczynski-NOAA replied:
Fixed

GwenChen-NOAA commented on jobs/rocoto/gempakgrb2spec.sh:
This task is for the JGFS_ATMOS_**GEMPAK_PGRB2_SPEC** job. The old name with "p" in the filename is fine.

WalterKolczynski-NOAA replied:
Gonna leave this for now. We can reconcile the difference between the j-job name and and rocoto job name later (probably by renaming it entirely).

WalterKolczynski-NOAA commented:
Think I've got the job-stopping bugs ironed out. Output is continuing to be produced on Cactus here:
`/lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak`. AWIPS and FSU Genesis are off.

@aerorahul is working on updating [NOAA-EMC/gfs-utils](https://github.com/NOAA-EMC/gfs-utils) to build the missing `rdbfmsua` executable.

WalterKolczynski-NOAA committed:
Fix typos in ukmet gempak meta

WalterKolczynski-NOAA committed:
Update gfs-utils for rdbfmsua build

Updates the gfs-utils build to enable building rdbfmsua. Also
uncomment the line calling rdbfmsua.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/fix_gempak

GwenChen-NOAA commented on gempak/ush/gfs_meta_hur.sh:
```diff
@@ -1,95 +1,62 @@
+        gdatpcpn12="F012-F126-06"
+        gdatpcpn24="F024-F126-06"
+        ;;
+esac
```
These 3 cases look the same to me. So do in the old code. Any reasons why gdatpcpn06, gdatpcpn48, gdatpcpn60, gdatpcpn72, gdatpcpn84, gdatpcpn96, gdatpcpn120, and gdatpcpn126 were taken out?

WalterKolczynski-NOAA replied:
The ones that were deleted weren't being used. I think you're right and the cyc doesn't really matter.

GwenChen-NOAA commented on gempak/ush/gfs_meta_sa2.sh:
```diff
@@ -34,30 +28,25 @@ device="nc | ${metaname}"
+PDYm1="$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} - 24 hours")"
 
-$GEMEXE/gdplot2_nc << EOF
+for symlink in ecwmf_m1 ukmet; do
```
This should be:
for symlink in ec**mw**f_m1 ukmet; do

WalterKolczynski-NOAA replied:
Fixed.

GwenChen-NOAA commented on gempak/ush/gfs_meta_sa2.sh:
```diff
@@ -362,26 +326,31 @@ HILO    = 31/H#;L#/1020-1060;900-1010/5/10;10!2/H#;L#/1020-1060;900-1010/5/10;10
 
-#export COMIN=/com/nawips/${envir}/ukmet.${PDY}
-echo " end with ukmet_gfs_meta_sa2.sh"
+mv "${metaname}" "${COM_ATMOS_GEMPAK_META}/${mdl}_${PDY}_${cyc}_${metatype}"
```
There is no gfs_20240314_00_sa2 file under /lfs/h1/ops/prod/com/gfs/v16.3/gfs.20240314/00/atmos/gempak/meta. So, similar to the gdas_ecmwf_meta_ver.sh and gdas_ukmet_meta_ver.sh scripts, this may be a leftover script that no longer needed.

GwenChen-NOAA replied:
This file is also generated for the 06 cycle only. It is located at /lfs/h1/ops/prod/com/gfs/v16.3/gfs.20240315/06/atmos/gempak/meta.

GwenChen-NOAA commented on gempak/ush/gfs_meta_us.sh:
```diff
@@ -158,24 +152,23 @@ export err=$?;err_chk
     fi
-  fi
 fi
 
```
Add `exit` here?

WalterKolczynski-NOAA replied:
Not necessary since this is the end of the script it implicitly does `exit 0`

GwenChen-NOAA commented on gempak/ush/gfs_meta_usext.sh:
```diff
@@ -240,21 +223,18 @@ export err=$?; err_chk
+        "${DBNROOT}/bin/dbn_alert" MODEL "${DBN_ALERT_TYPE}" "${job}" \
+            "${COM_ATMOS_GEMPAK_META}/gfs_${PDY}_${cyc}_usext"
+    fi
+fi
```
Add `exit` here?

WalterKolczynski-NOAA replied:
Same.

GwenChen-NOAA commented on gempak/ush/gfs_meta_ver.sh:
```diff
@@ -1,372 +1,65 @@
+#GENERATING THE METAFILES.
+# seq won't give us any splitting problems, ignore warnings
+# shellcheck disable=SC2207
+IFS=$'\n' lookbacks=($(seq 6 6 198))
```
I think this should be:
`IFS=$'\n' lookbacks=($(seq 6 6 180) $(seq 192 12 216))`

WalterKolczynski-NOAA replied:
Fixed

GwenChen-NOAA commented on gempak/ush/gfs_meta_comp.sh:
```diff
@@ -1,184 +1,153 @@
-export HPCNAM=${COMINnam}.$PDY/gempak
+export COMIN="gfs.multi"
+mkdir "${COMIN}"
+for cycle in $(seq -f "%02g" -s ' ' 0  "${STEP_GFS}" "${cyc}"); do
```
I don't see `${STEP_GFS}` defined prior to this line or in the j-job or ex-script. Is it defined in preamble.sh or elsewhere?

WalterKolczynski-NOAA replied:
`STEP_GFS` is defined in `config.base`

GwenChen-NOAA commented on gempak/ush/gfs_meta_mar_comp.sh:
```diff
@@ -1,158 +1,154 @@
 
-cd $DATA/MAR_COMP
-cp ${HOMEgfs}/gempak/fix/datatype.tbl datatype.tbl
+ource "${HOMEgfs}/ush/preamble.sh"
```
Missing "s". Should be:
`source "${HOMEgfs}/ush/preamble.sh"`

WalterKolczynski-NOAA replied:
Fixed

GwenChen-NOAA commented on jobs/JGFS_ATMOS_GEMPAK_NCDC_UPAPGIF:
```diff
@@ -28,37 +29,40 @@ export COMPONENT="atmos"
+done
+
+YMD=${PDY} HH=${cyc} generate_com -rx COM_ATMOS_WMO
+if [[ ! -d "${COM_ATMOS_WMO}" ]]; then mkdir m 775 -p "${COM_ATMOS_WMO}"; fi
```
Missing "-". Should be:
`mkdir -m 775 -p "${COM_ATMOS_WMO}"`

WalterKolczynski-NOAA replied:
Fixed

GwenChen-NOAA commented on scripts/exgfs_atmos_gempak_gif_ncdc_skew_t.sh:
```diff
@@ -7,107 +7,79 @@
+    sleep_interval=20
+    max_tries=180
+    for fhr in ${fcsthrs}; do
+        fhr3=$(printf %03i "${fhr}")
```
Should this be `fhr3=$(printf "%03d" "${fhr}")`?

WalterKolczynski-NOAA replied:
They're function almost identically, but fixed.

GwenChen-NOAA commented on scripts/exgfs_atmos_grib2_special_npoess.sh:
```diff
@@ -146,57 +133,44 @@ done
-   done
-   set_trace
+   grib_file="${COM_ATMOS_MASTER}/${RUN}.t${cyc}z.goesmasterf${fhr3}.grb2"
+   if ! wait_for_file "${grib_file}" "${SLEEP_TIME}" "${SLEEP_LOOP_MAX}"; then
```
`"${SLEEP_TIME}"` or `"${SLEEP_INT}"`?

WalterKolczynski-NOAA replied:
Should be `SLEEP_INT` like the one earlier in the script. Fixed.

GwenChen-NOAA commented on workflow/applications/gfs_cycled.py:
```diff
@@ -254,8 +256,9 @@ def get_task_names(self):
-            gfs_tasks += ['gempakpgrb2spec']
+            if self.do_goes:
+                gfs_tasks += ['npoess_pgrb2_0p5deg']
+                gfs_tasks += ['gempakpgrb2spec']
```
Does the task name `gempakpgrb2spec` need to be consistent with the one in rocoto/gfs_tasks.py? If so, it should be `gempakgrb2spec` without "p".

WalterKolczynski-NOAA replied:
All that matter is that is matches the function name in `gfs_tasks.py` (which is does). That function specifies the bash script to use as the job entrypoint.

WalterKolczynski-NOAA committed:
Fix typos discovered during code review

WalterKolczynski-NOAA committed:
Merge branch develop into feature/fix_gempak

WalterKolczynski-NOAA committed:
Turn rdbfmsua off

The build of rdbfmsua by gfs-utils caused conflicts with gfs_bufr, so
it had to be disabled for now (NOAA-EMC/gfs-utils#54). So also disable
the usage of the program in workflow.

WalterKolczynski-NOAA committed:
Update gfs_utils hash

aerorahul reviewed (approved):
I did not go through every script in detail, but the changes overall look good to me and consistent.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/fix_gempak

WalterKolczynski-NOAA committed:
Fix webtitle executable name for GDAS

GwenChen-NOAA commented:
For sample output GDAS gempak files from the test run at /lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak/gdas.20211231/00/products/atmos/gempak:

1. Cannot use gempak gdinfo to list out the parameters containing in the gempak files located at gdas.20211231/00/products/atmos/gempak/1p00 & 0p25. Please see an example output for the gdas_1p00_2021123100f003 file, comparing to one from /com (gdas_2024031700f003.txt) using the chkgempak.sh script.
2. The gdas_20211231_00_loop file is missing under gdas.20211231/00/products/atmos/gempak/meta.
3. No gif files are generated under gdas.20211231/00/products/atmos/gempak/gif.

[gdas_2024031700f003.txt](https://github.com/NOAA-EMC/global-workflow/files/14641647/gdas_2024031700f003.txt)
[gdas_1p00_2021123100f003.txt](https://github.com/NOAA-EMC/global-workflow/files/14641648/gdas_1p00_2021123100f003.txt)
[chkgempak.sh.txt](https://github.com/NOAA-EMC/global-workflow/files/14641649/chkgempak.sh.txt)

WalterKolczynski-NOAA committed:
Link rdbfmsua executable

WalterKolczynski-NOAA committed:
Fix some file paths in gdas meta loop

WalterKolczynski-NOAA commented:
> For sample output GDAS gempak files from the test run at /lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak/gdas.20211231/00/products/atmos/gempak:
> 
> 1. Cannot use gempak gdinfo to list out the parameters containing in the gempak files located at gdas.20211231/00/products/atmos/gempak/1p00 & 0p25. Please see an example output for the gdas_1p00_2021123100f003 file, comparing to one from /com (gdas_2024031700f003.txt) using the chkgempak.sh script.
> 2. The gdas_20211231_00_loop file is missing under gdas.20211231/00/products/atmos/gempak/meta.
> 3. No gif files are generated under gdas.20211231/00/products/atmos/gempak/gif.
> 
> [gdas_2024031700f003.txt](https://github.com/NOAA-EMC/global-workflow/files/14641647/gdas_2024031700f003.txt) [gdas_1p00_2021123100f003.txt](https://github.com/NOAA-EMC/global-workflow/files/14641648/gdas_1p00_2021123100f003.txt) [chkgempak.sh.txt](https://github.com/NOAA-EMC/global-workflow/files/14641649/chkgempak.sh.txt)

For (1), your filename is being truncated so gempak can't find it (look at the GDFILE in the log). If you use a shorter path (by going to the directory or through a symlink), it works:
```
 GEMPAK-GDINFO>l
 GDFILE   = gdas_1p00_2021123100f003
 LSTALL   = YES
 OUTPUT   = T
 GDATTIM  = all
 GLEVEL   = all
 GVCORD   = all
 GFUNC    = all
 GEMPAK-GDINFO>r

 GRID FILE: gdas_1p00_2021123100f003                                                                            

 GRID NAVIGATION: 
     PROJECTION:          CED                 
     ANGLES:                 0.0     0.0     0.0
     GRID SIZE:              360     181
     LL CORNER:           -90.0000    0.0000
     UR CORNER:            90.0000   -1.0000

 GRID ANALYSIS BLOCK: 
     ANALYSIS TYPE:        BARNES
     DELTAN:               4.000
     DELTAX:           -9999.000
     DELTAY:           -9999.000
     GRID AREA:           -90.00 -180.00   90.00  180.00
     EXTEND AREA:         -90.00 -180.00   90.00  180.00
     DATA AREA:           -90.00 -180.00   90.00  180.00

 Number of grids in file:   666

 Maximum number of grids in file:   4999

  NUM       TIME1              TIME2           LEVL1 LEVL2  VCORD PARM
    1     211231/0000F003                          0         TROP PRES        
    2     211231/0000F003                          0         TROP ICAHT       
    3     211231/0000F003                          0         TROP HGHT        
    4     211231/0000F003                          0         TROP TMPK        
    5     211231/0000F003                          0         TROP UREL        
    6     211231/0000F003                          0         TROP VREL        
    7     211231/0000F003                          0         TROP VWSH        
    8     211231/0000F003                          0         FRZL HGHT        
    9     211231/0000F003                          0         FRZL RELH        
   10     211231/0000F003                          0         HTFL HGHT        
   11     211231/0000F003                          0         HTFL RELH        
   12     211231/0000F003                          0         MCBL PRES03      
   13     211231/0000F003                          0         LCBL PRES03      
   14     211231/0000F003                          0         HCBL PRES03      
   15     211231/0000F003                          0         CCBL PRES        
   16     211231/0000F003                          0         CEIL HGHT        
   17     211231/0000F003                          1         HYBL CWTR        
   18     211231/0000F003                          1         HYBL ICMR        
   19     211231/0000F003                          1         HYBL RNMX        
   20     211231/0000F003                          1         HYBL SNMR        
   21     211231/0000F003                          1         HYBL GRMR        
   22     211231/0000F003                          0  10     DPTH TMPK        
   23     211231/0000F003                          0  10     DPTH SOILM       
   24     211231/0000F003                          0  10     DPTH SOILL       
   25     211231/0000F003                        100  200    DPTH TMPK        
   26     211231/0000F003                        100  200    DPTH SOILM       
   27     211231/0000F003                        100  200    DPTH SOILL       
   28     211231/0000F003                         40  100    DPTH TMPK        
   29     211231/0000F003                         40  100    DPTH SOILM       
   30     211231/0000F003                         40  100    DPTH SOILL       
   31     211231/0000F003                         10  40     DPTH TMPK        
   32     211231/0000F003                         10  40     DPTH SOILM       
   33     211231/0000F003                         10  40     DPTH SOILL       
   34     211231/0000F003                       9950         SGMA TMPK        
   35     211231/0000F003                       9950         SGMA THTA        
Enter <CR> to page or type EXIT:
```

WalterKolczynski-NOAA commented:
For (2), I've made some fixes and am running a second experiment (change the experiment name from `gempak` to `gempak_2` in the path). This experiment looks to be generating loop files (fixes will be committed shortly):
```
WCOSS2 (BACKUPSYS) meta> pwd
/lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak_2/gdas.20211222/00/products/atmos/gempak/meta
WCOSS2 (BACKUPSYS) meta> ls
gdas_20211222_00_loop  gdas_20211222_00_na
```

WalterKolczynski-NOAA commented:
(3) I'm still working on, but maybe we'll merge what is here and then fix that in a follow-up.

WalterKolczynski-NOAA committed:
Fix issue with ukmet symlink

WalterKolczynski-NOAA committed:
Fix gdas gempak gif

WalterKolczynski-NOAA commented:
(3) has been solved. gempak_2 is producing the gdas gifs now:

```
/lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak_2/gdas.20211224/00/products/atmos/gempak/gif:
total 1.9M
-rw-r--r-- 1 walter.kolczynski global 192K Mar 20 06:54 gdas_200_hgt_iso_nh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 201K Mar 20 06:54 gdas_250_hgt_iso_nh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 178K Mar 20 06:54 gdas_250_hgt_iso_sh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 201K Mar 20 06:54 gdas_300_hgt_iso_nh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 171K Mar 20 06:54 gdas_300_hgt_iso_sh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 149K Mar 20 06:54 gdas_500_hgt_tmp_nh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 135K Mar 20 06:54 gdas_500_hgt_tmp_sh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 159K Mar 20 06:54 gdas_700_hgt_tmp_nh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 157K Mar 20 06:54 gdas_850_hgt_tmp_nh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 172K Mar 20 06:54 gdas_sfc_mslp_thk_nh_anl_00.gif
-rw-r--r-- 1 walter.kolczynski global 152K Mar 20 06:54 gdas_sfc_mslp_thk_sh_anl_00.gif
```

Also made it so the gif directory is only created for the cycles that create gifs (00 and 12).

GwenChen-NOAA commented:
> > For sample output GDAS gempak files from the test run at /lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak/gdas.20211231/00/products/atmos/gempak:
> > 
> > 1. Cannot use gempak gdinfo to list out the parameters containing in the gempak files located at gdas.20211231/00/products/atmos/gempak/1p00 & 0p25. Please see an example output for the gdas_1p00_2021123100f003 file, comparing to one from /com (gdas_2024031700f003.txt) using the chkgempak.sh script.
> > 2. The gdas_20211231_00_loop file is missing under gdas.20211231/00/products/atmos/gempak/meta.
> > 3. No gif files are generated under gdas.20211231/00/products/atmos/gempak/gif.
> > 
> 
> For (1), your filename is being truncated so gempak can't find it (look at the GDFILE in the log). If you use a shorter path (by going to the directory or through a symlink), it works.
> 

Yes, you are right. Glad to know this is a limitation of gempak software, not related to the code.

GwenChen-NOAA commented:
> For (2), I've made some fixes and am running a second experiment (change the experiment name from `gempak` to `gempak_2` in the path). This experiment looks to be generating loop files (fixes will be committed shortly):
> 
> ```
> WCOSS2 (BACKUPSYS) meta> pwd
> /lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak_2/gdas.20211222/00/products/atmos/gempak/meta
> WCOSS2 (BACKUPSYS) meta> ls
> gdas_20211222_00_loop  gdas_20211222_00_na
> ```

Thanks, @WalterKolczynski-NOAA!  I see the gdas_20211222_00_loop file in the new run now.  The ukmetver_20211222_12 and ecmwfver_20211222_12 files are missing at /lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak_2/gdas.20211222/06/products/atmos/gempak/meta. These two files are only generated for the 06 cycle. In /com, they are located at:

/lfs/h1/ops/prod/com/gfs/v16.3/gdas.20240319/06/atmos/gempak/meta/ukmet.20240319/meta/ukmetver_20240319_12
/lfs/h1/ops/prod/com/gfs/v16.3/gdas.20240319/06/atmos/gempak/meta.20240319/meta/ecmwfver_20240319_12

WalterKolczynski-NOAA committed:
Turn rdbfmsua back on

A gfs-utils update allows the compilation of rdbfmsua without causing
library issues with bufr.

GwenChen-NOAA commented:
@WalterKolczynski-NOAA, could you tell me how you set up the ecmwf, ukmet, and nam data to run this PR?  Or point me to where the input data are located on WCOSS2?  I cloned your fork and checked out the feature/fix_gempak branch, but I cannot run the gempak meta jobs because they require input data from ecmwf, ukmet, and nam.  I think this PR will be ready to be merged once the ukmetver_20211222_12 and ecmwfver_20211222_12 files can be generated.  Thanks!

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA, could you tell me how you set up the ecmwf, ukmet, and nam data to run this PR? Or point me to where the input data are located on WCOSS2? I cloned your fork and checked out the feature/fix_gempak branch, but I cannot run the gempak meta jobs because they require input data from ecmwf, ukmet, and nam. I think this PR will be ready to be merged once the ukmetver_20211222_12 and ecmwfver_20211222_12 files can be generated. Thanks!

If you clone and setup an experiment from my branch, you should have `COMINecmwf`, `COMINnam`, and 
`COMINukmet` set in your `config.base`:

```
export COMINecmwf=/lfs/h2/emc/global/noscrub/emc.global/data/external_gempak/ecmwf
export COMINnam=/lfs/h2/emc/global/noscrub/emc.global/data/external_gempak/nam
export COMINukmet=/lfs/h2/emc/global/noscrub/emc.global/data/external_gempak/ukmet
```

WalterKolczynski-NOAA commented:
> The ukmetver_20211222_12 and ecmwfver_20211222_12 files are missing at /lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak_2/gdas.20211222/06/products/atmos/gempak/meta. These two files are only generated for the 06 cycle. In /com, they are located at:
> 
> /lfs/h1/ops/prod/com/gfs/v16.3/gdas.20240319/06/atmos/gempak/meta/ukmet.20240319/meta/ukmetver_20240319_12 /lfs/h1/ops/prod/com/gfs/v16.3/gdas.20240319/06/atmos/gempak/meta.20240319/meta/ecmwfver_20240319_12

I'll look into this Monday when I am back from leave.

WalterKolczynski-NOAA committed:
Only create gif directory when needed

WalterKolczynski-NOAA committed:
More symlinks for gempak

WalterKolczynski-NOAA committed:
Standardize gempak symlinks

WalterKolczynski-NOAA committed:
Add needed export and fix bug in gempak

WalterKolczynski-NOAA committed:
Delay cleanup when gempak is on

Gempak looks back up to 216h, so don't clean cycles as quickly when
gempak is run.

WalterKolczynski-NOAA committed:
One more gempak fix

WalterKolczynski-NOAA committed:
Fix error in cleanup config

WalterKolczynski-NOAA commented:
@GwenChen-NOAA I went through all the scripts and made sure they were all using the symlinks to avoid the path length problem. The `ecmwfver_${PDY}_12` and `ukmet_${PDY}_12` files are now being produced:
```
WCOSS2 (BACKUPSYS) meta> pwd
/lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak_2/gdas.20211229/06/products/atmos/gempak/meta
WCOSS2 (BACKUPSYS) meta> ls -l
total 26M
-rw-r--r-- 1 walter.kolczynski emc 2.1M Mar 27 00:00 ecmwfver_20211229_12
-rw-r--r-- 1 walter.kolczynski emc 7.4M Mar 27 00:00 gdas_20211229_06_loop
-rw-r--r-- 1 walter.kolczynski emc  14M Mar 27 00:00 gdas_20211229_06_na
-rw-r--r-- 1 walter.kolczynski emc 2.7M Mar 27 00:00 ukmetver_20211229_12
```

After the production switch, I will set up a clean run on Dogwood as a final check.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/fix_gempak

WalterKolczynski-NOAA committed:
Use correct variable for SA2

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/fix_gempak

GwenChen-NOAA commented:
> @GwenChen-NOAA I went through all the scripts and made sure they were all using the symlinks to avoid the path length problem. The `ecmwfver_${PDY}_12` and `ukmet_${PDY}_12` files are now being produced:
> 
> ```
> WCOSS2 (BACKUPSYS) meta> pwd
> /lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak_2/gdas.20211229/06/products/atmos/gempak/meta
> WCOSS2 (BACKUPSYS) meta> ls -l
> total 26M
> -rw-r--r-- 1 walter.kolczynski emc 2.1M Mar 27 00:00 ecmwfver_20211229_12
> -rw-r--r-- 1 walter.kolczynski emc 7.4M Mar 27 00:00 gdas_20211229_06_loop
> -rw-r--r-- 1 walter.kolczynski emc  14M Mar 27 00:00 gdas_20211229_06_na
> -rw-r--r-- 1 walter.kolczynski emc 2.7M Mar 27 00:00 ukmetver_20211229_12
> ```
> 
> After the production switch, I will set up a clean run on Dogwood as a final check.

Thanks, @WalterKolczynski-NOAA! Once you have it running on Dogwood, let me know. I can help check the outputs. I'm working with NCO SDB to access NAWIPS software, so I can view the metafiles. It may take some time. We can address in another PR if there is any issue after viewing the metafiles. Other files (gempak data files, upper air ASCII files, and image files) all look good to me.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/fix_gempak

WalterKolczynski-NOAA commented:
@GwenChen-NOAA Run is populating in `/lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak` on Dogwood

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/fix_gempak

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/fix_gempak

WalterKolczynski-NOAA commented:
Running CI for completeness, even though it will not test the gempak jobs.

GwenChen-NOAA commented:
@WalterKolczynski-NOAA, all gempak products under your test run in /lfs/h2/emc/global/noscrub/walter.kolczynski/global-workflow/gempak look good to me. I will approve this PR shortly. Thanks for the tremendous efforts you put into this PR to get the gempak jobs running!

CatherineThomas-NOAA commented:
@aerorahul @DavidHuber-NOAA - Do either of you remember why [this](https://github.com/NOAA-EMC/global-workflow/blame/9c9abbd7ea43a98ca7c34c2390c8ef4166a2400b/dev/parm/config/gfs/config.cleanup#L15) was added to `config.cleanup` in this PR?
```
if [[ "${DO_GEMPAK}" == "YES" ]]; then
    export RMOLDSTD=346
    export RMOLDEND=222
fi
```

Edited to add some context:  We're running with GEMPAK in the v17 real time run and found that this is why our COM hasn't been scrubbing as expected.

DavidHuber-NOAA commented:
@WalterKolczynski-NOAA Any ideas? ^^

DavidHuber-NOAA commented:
@CatherineThomas-NOAA It looks like this came in to prevent cleaning up too much data when gempak was still running, based on this commit: https://github.com/NOAA-EMC/global-workflow/pull/2374/commits/ab4bc385998e071ee4645b8904494d6ba37c1e21

Can you share what isn't being scrubbed that should be?

CatherineThomas-NOAA commented:
@DavidHuber-NOAA - Good find on that commit, it looks pretty clear there.

As for what data isn't being scrubbed, it's all of it.  The cleanup step isn't scrubbing anything until it reaches 222 hours, including the ensemble data (which by itself is 11TB per cycle for full res).  

We selectively exclude certain files from cleaning already.  Maybe we could use this for GEMPAK or make use of the vrfyarch directory?
